akka-stream-tests/src/test/scala/akka/stream/scaladsl/MapWithResourceSpec.scala

```diff
@@ -0,0 +1,262 @@
+      val p = Source
+        .repeat(1)
+        .take(100)
+        .mapWithResource(() => newBufferedReader())((reader, _) => {
+          val s = reader.readLine()
+          if (s != null && s.contains("b")) throw TE("") else s
+        }, reader => reader.close())
+        .withAttributes(supervisionStrategy(resumingDecider))
+        .runWith(Sink.asPublisher(false))
+      val c = TestSubscriber.manualProbe[String]()
+
+      p.subscribe(c)
+      val sub = c.expectSubscription()
+
+      (0 to 49).foreach(i => {
+        sub.request(1)
+        c.expectNext() should ===(if (i < 10) manyLinesArray(i) else manyLinesArray(i + 10))
+      })
+      sub.request(1)
+      c.expectComplete()
+    }
+
+    "close and open stream again when Strategy is Restart" in {
+      val p = Source
+        .repeat(1)
+        .take(100)
+        .mapWithResource(() => newBufferedReader())((reader, _) => {
+          val s = reader.readLine()
+          if (s != null && s.contains("b")) throw TE("") else s
+        }, reader => reader.close())
+        .withAttributes(supervisionStrategy(restartingDecider))
+        .runWith(Sink.asPublisher(false))
+      val c = TestSubscriber.manualProbe[String]()
+
+      p.subscribe(c)
+      val sub = c.expectSubscription()
+
+      (0 to 19).foreach(_ => {
+        sub.request(1)
+        c.expectNext() should ===(manyLinesArray(0))
+      })
+      sub.cancel()
+    }
+
+    "work with ByteString as well" in {
+      val chunkSize = 50
+      val buffer = new Array[Char](chunkSize)
+      val p = Source
+        .repeat(1)
+        .mapWithResource(() => newBufferedReader())((reader, _) => {
+          val s = reader.read(buffer)
+          if (s > 0) Some(ByteString(buffer.mkString("")).take(s)) else None
+        }, reader => reader.close())
+        .takeWhile(_.isDefined)
+        .collect {
+          case Some(bytes) => bytes
+        }
+        .runWith(Sink.asPublisher(false))
+      val c = TestSubscriber.manualProbe[ByteString]()
+
+      var remaining = manyLines
+      def nextChunk() = {
+        val (chunk, rest) = remaining.splitAt(chunkSize)
+        remaining = rest
+        chunk
+      }
+
+      p.subscribe(c)
+      val sub = c.expectSubscription()
+
+      (0 to 121).foreach(_ => {
+        sub.request(1)
+        c.expectNext().utf8String should ===(nextChunk().toString)
+      })
+      sub.request(1)
+      c.expectComplete()
+    }
+
+    "use dedicated blocking-io-dispatcher by default" in {
+      val p = Source
+        .single(1)
+        .mapWithResource(() => newBufferedReader())((reader, _) => Option(reader.readLine()), reader => reader.close())
+        .runWith(TestSink.probe)
+
+      SystemMaterializer(system).materializer
+        .asInstanceOf[PhasedFusingActorMaterializer]
+        .supervisor
+        .tell(StreamSupervisor.GetChildren, testActor)
+      val ref = expectMsgType[Children].children.find(_.path.toString contains "statefulMap").get
```

Comment:
I have no idea why the name of `mapWithResource` is not take effect, seems like a bug?
![lQLPJxand6_Y7QHNBUzNC8Sw8phxE7O8NjEDE3HxscCEAA_3012_1356](https://user-images.githubusercontent.com/501740/188304245-c5080c71-96b2-4a97-87de-2390a8562db9.png)

Reply:
May be a bug, but names are also a bit special and doesn't quite fit with the attribute design where closest defined attribute of a type "wins". 

I can't find any stage that is both named on GraphStage level _and_ factory methods, the easiest solution may be to move the name of `statefulMap` out to the factories so that there is no conflict, the `StatefulMap` class is internal so no risk it is used any other way than through the factory methods.